src/cmd/services/m3coordinator/downsample/metrics_appender.go

```diff
@@ -117,10 +124,29 @@ type metricsAppenderOptions struct {
 	metrics      metricsAppenderMetrics
 }
 
-func newMetricsAppender(pool *metricsAppenderPool) *metricsAppender {
+func newMetricsAppender(
+	pool *metricsAppenderPool,
+	tagIterPool serialize.MetricTagsIteratorPool,
+	nameTag []byte) *metricsAppender {
+	tagIter := tagIterPool.Get()
 	return &metricsAppender{
 		pool:                 pool,
 		multiSamplesAppender: newMultiSamplesAppender(),
+		tagIterFn: func(tagPairs []byte) id.SortedTagIterator {
+			// Use the same tagIter for all matching computations. It is safe to reuse since a metric appender is
+			// single threaded and it's reset before each computation.
+			tagIter.Reset(tagPairs)
+			return tagIter
+		},
+		nameTagFn: func(id []byte) ([]byte, []byte, error) {
```

Comment:
it wasn't strictly necessary to move this fn to be request scoped, but I think it provides consistency and we can probably remove the `TagValueFromEncodedTagsFast` in favor of just reusing a tag decoder like a tag iterator. I'm sure we added this for a similar pooling issue lol.